Readme and JSDoc proofreading done

praveen5959 opened the pull request:
Summary

READEME and JSDoc comments in the code have been proofread.

praveen5959 committed:
Readme and JSDoc proofreading done

praveen5959 committed:
Lint issues fixed

balazsbotond commented:
Thank you for taking the time to proofread everything! Could you please also explain *why* you proposed each change?

praveen5959 commented:
In the index.ts file, some JSDoc comments didna have appropriate fields like return value so I have added those.
For some function, JSDoc comments were not present so I have added those as well.

In README file changes I have done to make the docs general and just first-person English.  And small other changes here and there.

balazsbotond commented on .gitignore:
```diff
@@ -3,3 +3,4 @@
 /dist-test
 /.nyc_output
 /coverage
+.vscode
```
Thanks for including this 👍

balazsbotond commented on README.md:
```diff
@@ -52,7 +52,7 @@ Features:
 ## Why?
 
-When I need to call an HTTP API, I usually need to add dynamic parameters to the URL:
+When we call an HTTP API, we usually need to add dynamic parameters to the URL:
```
I wrote this section in the first person because I want the README to be more personal. To be honest I would want to keep it that way; still, thanks for taking the time to read through it.

balazsbotond commented on README.md:
```diff
@@ -63,12 +63,12 @@ function getUserPosts(id, blogId, limit, offset) {
+As we can see, this minimal example is already rather hard to read. It is also incorrect:
 
-- I forgot that there was a trailing slash at the end of the `API_URL` constant so the slash got duplicated (`https://api.example.com//users`)
+- One common issue is of trailing slashes, `API_URL` has a trailing slash in the example. This will lead to URL containing duplicated slashes (`https://api.example.com//users`).
```
I agree that the phrase "the slash got duplicated" sounds unnatural. Can we just keep the second part ("This will lead to URL containing duplicated slashes") from here?

balazsbotond commented on README.md:
```diff
@@ -63,12 +63,12 @@ function getUserPosts(id, blogId, limit, offset) {
 - The embedded values need to be escaped using `encodeURIComponent`
 
-I can use the built-in `URL` class to prevent duplicate slashes and `URLSearchParams` to escape the query string. But I still need to escape all path parameters manually.
+We can use the built-in `URL` class to prevent duplicate slashes and `URLSearchParams` to escape the query string. But we still need to escape all path parameters manually.
```
As I said earlier, I would like to keep the first person voice. Could you please undo the changes related to it?

balazsbotond commented on README.md:
```diff
@@ -103,29 +103,30 @@ The library handles:
+Since the code uses the `URL` and `URLSearchParams` classes internally, which aren't available below v10, we cannot support those versions.
 
-If you want to build full URLs (most common use case):
+To build full URLs (most common use case):
```
I agree with these 👍 Let's keep them

balazsbotond commented on README.md:
````diff
@@ -103,29 +103,30 @@ The library handles:
 ```
 
-If you want to use everything:
+To use everything:
````
Let's change this to "To use all exported functions:"

balazsbotond commented on README.md:
```diff
@@ -262,7 +264,7 @@ Joins the two parts using exactly one separator. If a separator is present at th
 Thank you for using *urlcat*!
 
-If you need any help using this library, feel free to [create a GitHub issue](https://github.com/balazsbotond/urlcat/issues/new/choose) and ask your questions. I'll try to answer as quickly as possible.
+For any kind of help needed while using this library, feel free to [create a GitHub issue](https://github.com/balazsbotond/urlcat/issues/new/choose), and ask your questions. I'll try to answer as quickly as possible.
```
I think we should keep the second person voice here. It's more personal and directly addresses the reader.

balazsbotond commented on README.md:
```diff
@@ -272,7 +274,8 @@ Contributions of any kind (pull requests, bug reports, feature requests, documen
 
-You need to have a reasonably recent version of node.js to build *urlcat*. My node version is 12.18.3, npm is at 6.14.6.
+You need to have a reasonably recent version of node.js to build *urlcat*. 
+Tested on node version 12.18.3 and npm version 6.14.6.
```
Thanks, this is much better this way.

balazsbotond commented on src/index.ts:
```diff
@@ -6,7 +6,8 @@ export type ParamMap = Record<string, any>;
  *   in the base template. Unused properties become query params.
- * @returns a URL with path params substituted and query params appended
+ *
+ * @returns {String} a URL with path params substituted and query params appended
```
👍

balazsbotond commented on src/index.ts:
```diff
@@ -17,12 +18,13 @@ export type ParamMap = Record<string, any>;
+ * If a '/' occurs at the concatenation boundary in either parameter, it is removed.
+ *
+ * @param {String} baseUrl Base URL
+ * @param {String} path Path of URL or query :params
```
I wrote "first part" and "second part" here because I wanted to emphasize that you can break up the URL anywhere. The updated description of the `path` parameter isn't correct, query params aren't handled by this overload.

balazsbotond commented on src/index.ts:
```diff
@@ -17,12 +18,13 @@ export type ParamMap = Record<string, any>;
- * @param {String} baseUrl the first part of the URL
- * @param {String} path the second part of the URL
- * @returns {String} the result of the concatenation
+ * @returns {String} Concatenated URL
```
👍

balazsbotond commented on src/index.ts:
```diff
@@ -33,16 +35,17 @@ export default function urlcat(baseTemplate: string, params: ParamMap): string;
- * Concatenates the base URL and the path specified using '/' as a separator such
- * that if a '/' occurs at the concatenation boundary in either parameter, it is removed.
+ * Concatenates the base URL and the path specified using '/' as a separator.
+ * If a '/' occurs at the concatenation boundary in either parameter, it is removed.
```
Thanks, sounds much better this way!

balazsbotond commented on src/index.ts:
```diff
@@ -33,16 +35,17 @@ export default function urlcat(baseTemplate: string, params: ParamMap): string;
- * @param {String} path the second part of the URL
- * @param {Object} params an object with properties that correspond to the :params
+ * @param {String} baseUrl Base URL
+ * @param {String} path Path of URL or query :params
```
Again, no query params allowed here

balazsbotond commented on src/index.ts:
```diff
@@ -79,7 +82,8 @@ function urlcatImpl(pathTemplate: string, params: ParamMap, baseUrl?: string) {
  * @param {Object} params an object to convert into a query string.
- * @returns {String} a query string.
+ *
+ * @returns {String} Query string.
```
👍

balazsbotond commented on src/index.ts:
```diff
@@ -96,7 +100,8 @@ export function query(params: ParamMap): string {
  * @param {Object} params on object with keys that correspond to the params in the template.
- * @returns {String}
+ *
+ * @returns {String} Rendering path after substitution .
```
I would change this to "Rendered path after substitution."

balazsbotond commented on src/index.ts:
```diff
@@ -109,6 +114,13 @@ export function subst(template: string, params: ParamMap): string {
+ * @param {Object} params on object with keys that correspond to the params in the template.
+ *
+ * @returns {String, String} Rendering path and rendering :params.
+ */
```
This is an internal function that does not need a jsdoc string. Can you please remove it?

balazsbotond commented on src/index.ts:
```diff
@@ -133,13 +145,14 @@ function path(template: string, params: ParamMap) {
- * boundary in either of the strings, it is removed. This prevents accidental duplication
- * of the separator.
+ * Joins two strings using a separator.
+ * If the separator occurs at the concatenation boundary in either parameter, it is removed.
```
"in either parameter" is incorrect; there are three parameters. I would keep my original version here.

balazsbotond commented on src/index.ts:
```diff
@@ -133,13 +145,14 @@ function path(template: string, params: ParamMap) {
- * @param {String} part2 the second string to join.
+ * @param {String} part1 First string.
+ * @param {String} separator Separator used for joining.
+ * @param {String} part2 Second string.
```
👍

balazsbotond commented on src/index.ts:
```diff
@@ -133,13 +145,14 @@ function path(template: string, params: ParamMap) {
+ * @param {String} separator Separator used for joining.
+ * @param {String} part2 Second string.
+ *
+ * @returns {String} Joined string.
```
👍

balazsbotond commented on src/index.ts:
```diff
@@ -148,13 +161,21 @@ function path(template: string, params: ParamMap) {
+  const p2 = part2.startsWith(separator)
+    ? part2.slice(separator.length)
+    : part2;
+  return p1 === "" || p2 === "" ? p1 + p2 : p1 + separator + p2;
```
I see why you reformatted the first two lines - they are slightly easier to read this way - but I see no reason to put the return statement on a single line. It's become very hard to read. Could you please format it again?

balazsbotond commented on src/index.ts:
```diff
@@ -148,13 +161,21 @@ function path(template: string, params: ParamMap) {
+ * @param params Object consisting of :params
+ *
+ * @returns {Object} Object of :params after removing Null and undefined values
+ */
```
This is an internal function that isn't exported so it doesn't need a jsdoc string.

balazsbotond commented on src/index.ts:
```diff
@@ -164,6 +185,13 @@ function removeNullOrUndef(params: ParamMap) {
+ *
+ * @returns {Boolean} true -> :param is not Null or not undefined
+ * @returns {Boolean} false -> :param is Null or undefined
+ */
```
This is an internal function that isn't exported so it doesn't need a jsdoc string.

balazsbotond reviewed (changes requested):
Thank you again for reading through everything and suggesting changes. I agree with many of your suggestions. I commented on each change and indicated whether I want to keep them. Would you please update the PR accordingly?

praveen5959 commented:
Sure
I will do it by today EOD.

balazsbotond commented:
> Sure
> I will do it by today EOD.

Thank you!

praveen5959 committed:
PR comments fixed

praveen5959 commented:
I have updated the PR with the requested changes.

balazsbotond commented:
Thank you! Merging 👍

balazsbotond commented:
@all-contributors please add @praveen5959 for docs
